execution/src/execution.cc

```diff
@@ -218,7 +220,68 @@ namespace leatherman { namespace execution {
         auto actual_options = options;
         function<bool(string&)> stderr_callback;
         setup_execute(stderr_callback, actual_options);
-        return execute(file, &arguments, &input, &environment, nullptr, stderr_callback, actual_options, timeout);
+        return execute(file, &arguments, &input, &environment, nullptr, nullptr, stderr_callback, actual_options, timeout);
+    }
+
+    result execute(
+        string const& file,
+        vector<string> const& arguments,
+        string const& input,
+        map<string, string> const& environment,
+        std::function<void(size_t)> pid_callback,
+        uint32_t timeout,
+        option_set<execution_options> const& options)
+    {
+        auto actual_options = options;
+        function<bool(string&)> stderr_callback;
+        setup_execute(stderr_callback, actual_options);
+        return execute(file, &arguments, &input, &environment, pid_callback, nullptr, stderr_callback, actual_options, timeout);
+    }
+
+    result execute(
+        std::string const& file,
+        std::vector<std::string> const& arguments,
+        std::string const& input,
+        std::string const& out_file,
+        std::string const& err_file,
+        std::map<std::string, std::string> const& environment,
+        std::function<void(size_t)> pid_callback,
+        uint32_t timeout,
+        lth_util::option_set<execution_options> const& options)
+    {
+        auto actual_options = options;
+        function<bool(string&)> stderr_callback;
+        function<bool(string&)> stdout_callback;
+        boost::nowide::ofstream out_stream;
+        boost::nowide::ofstream err_stream;
+
+        out_stream.open(out_file.c_str(), std::ios::binary);
+        if (!out_stream.is_open()) {
+            throw execution_exception("failed to open the output file.");
+        }
+
+        if (err_file.empty()) {
+            setup_execute(stderr_callback, actual_options);
+        } else {
+            err_stream.open(err_file.c_str(), std::ios::binary);
+            if (!err_stream.is_open()) {
+                throw execution_exception("failed to open the error file.");
+            }
+
+            stderr_callback = ([&](string& line) {
+                err_stream << line << std::endl;
```

Comment:
Yeah, using endl was on purpose when I wrote it.
But on second thought, it's not worth.
I will change and squash the two commits.